alf/layers.py

```diff
@@ -1092,3 +1092,214 @@ def calc_output_shape(self, input_shape):
+            relative_positional_encoding (bool): If True, use the relative positional
+                encoding proposed in [3]. If False, use absolute positional encoding.
+        """
+        super().__init__()
+        if d_k is None:
+            d_k = d_model
+        if d_v is None:
+            d_v = d_model
+        self._q_proj = nn.Parameter(torch.Tensor(d_model, num_heads * d_k))
+        self._k_proj = nn.Parameter(torch.Tensor(d_model, num_heads * d_k))
+        self._v_proj = nn.Parameter(torch.Tensor(d_model, num_heads * d_v))
+        self._o_proj = nn.Parameter(torch.Tensor(num_heads * d_v, d_model))
+
+        self._d_model = d_model
+        self._d_k = d_k
+        self._d_v = d_v
+        self._num_heads = num_heads
+        self._memory_size = memory_size
+        self._relative_positional_encoding = relative_positional_encoding
+
+        self._attention_scale = 1. / d_k**0.5
+        self._mlp = torch.nn.Sequential(
+            FC(d_model, d_model, torch.relu_), FC(d_model, d_model))
+        self._norm1 = torch.nn.LayerNorm(d_model)
+        self._norm2 = torch.nn.LayerNorm(d_model)
+
+        l = 2 * memory_size - 1 if relative_positional_encoding else memory_size
+        self._positional_encoding = nn.Parameter(torch.Tensor(d_k, l))
+        self._qk_bias = nn.Parameter(torch.Tensor(1, num_heads, 1, d_k))
+        self._qp_bias = nn.Parameter(torch.Tensor(1, num_heads, 1, d_k))
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self._q_proj)
+        nn.init.xavier_uniform_(self._k_proj)
+        nn.init.xavier_uniform_(self._v_proj)
+        nn.init.xavier_uniform_(self._o_proj)
+        nn.init.zeros_(self._qk_bias)
+        nn.init.zeros_(self._qp_bias)
+        nn.init.zeros_(self._positional_encoding)
+        for l in self._mlp:
+            l.reset_parameters()
+
+    @staticmethod
+    def _shift(x):
+        """
+        y[:, :, i, j] <- x[:, :, i, n - 1 + i - j]
+        Args:
+            x: [B, H, M, 2 * N - 1]
+        Returns:
+            [B, H, M, N]
+        """
+        b, h, m, l = x.shape
+        n = (l + 1) // 2
+        # [M, N], index[i, j] = i * l + n  - 1 + i - j = n - 1 + (l + 1) * i - j
+        index = n - 1 + (l + 1) * torch.arange(m).unsqueeze(-1) - torch.arange(
+            n)
+        # [M * N]
+        index = index.reshape(-1)
+        # [M * L, B * H]
+        x_flattened = x.reshape(b * h, m * l).transpose(0, 1)
+        # [M * N, B * H]
+        y = x_flattened[index]
+        y = y.transpose(0, 1).reshape(b, h, m, n)
+        return y
+
+    def forward(self, memory, query=None, mask=None):
+        """
+        Args:
+            memory (Tensor): The shape is [B, N, d_model]
+            query (Tensor): The shape [B, d_model] or [B, M, d_model]. If None,
+                will use memory as query
+            mask (Tensor|None): [B, N] or [B, M, N]
```

Comment:
Is mask used to exclude certain memory positions? Maybe can add comments explaining the purpose of mask or modify the name?